### PR TITLE
Add semantic-segmentation model HRNetV2

### PR DESCRIPTION
**Pull request by alexmasterdi**
At this moment model based on opset11 from onnx.

### CODE REVIEW

File: models/public/hrnetv2/hrnetv2.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# hrnetv2
```
Comment (eizamaliev) [inline models/public/hrnetv2/hrnetv2.md:1]: I'll recommend to rename model to `hrnet-v2` or even `hrnet-v2-w48` if `w48` suffix is quite meaningful.

**Review by eizamaliev — CHANGES_REQUESTED** [review]

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

File: models/public/hrnetv2/hrnetv2.md (hunk 1 above)
Comment (eaidova) [reply models/public/hrnetv2/hrnetv2.md:1]: w48 is model size, for semantic segmentation there is no any other available (small models w18 and w32 used as backbone for facial landmarks recognition and pose estimation respectively, w48 is the most popular as I know). So it can be ommited, I think. But there is another important thing which should be reflected in the name, HRNetV2 is only half of this model, its encoder part. @alexmasterdi, please add info about decoder into model name

File: models/public/hrnetv2/accuracy-check.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
+models:
+
```
Comment (eaidova) [inline models/public/hrnetv2/accuracy-check.yaml:2]: please remove trailing line

File: models/public/hrnetv2/accuracy-check.yaml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+  - name: hrnetv2_pytorch
+
+    launchers:
+      - framework: dlsdk
+        device: CPU
+        model:   public/hrnetv2/hrnetv2.onnx
```
Comment (eaidova) [inline models/public/hrnetv2/accuracy-check.yaml:9]: if you want run onnx model, please use onnx_runtime as launcher.
Our current config formatting is not include running models via onnx importer, I think we add it later for all models if it will be necessary.
Please remove this section or change launcher on onnx_runtime

File: models/public/hrnetv2/accuracy-check.yaml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        model:   public/hrnetv2/FP16/hrnetv2.xml
+        weights: public/hrnetv2/FP16/hrnetv2.bin
+        adapter: segmentation
+
+      - framework: dlsdk
+        device: CPU
+        tags:
+          - FP32
+        model:   public/hrnetv2/FP32/hrnetv2.xml
+        weights: public/hrnetv2/FP32/hrnetv2.bin
```
Comment (eaidova) [inline models/public/hrnetv2/accuracy-check.yaml:53]: please remove device, tags and path to the models
```suggestion
```

File: models/public/hrnetv2/accuracy-check.yaml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        adapter: segmentation
+ 
+    datasets:
+      - name: ADEChallengeData2016
+        preprocessing:
+          - type: bgr_to_rgb
```
Comment (eaidova) [inline models/public/hrnetv2/accuracy-check.yaml:59]: why do you need bgr_to_rgb?
is it impossible to include --reverse_input_channels to the model during MO conversion?

File: models/public/hrnetv2/hrnetv2.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+## Use Case and High-Level Description
+HRNetV2-W48 is semantic-segmentation model based on architecture described in paper
+[High-Resolution Representations for Labeling Pixels and Regions](https://arxiv.org/abs/1904.04514).
+This is PyTorch implementation based on retaining high resolution representations throughout the model
+and pretrained on ADE20k dataset. The model input is a blob that consists of a
+single image of "1x3x320x320" in RGB order.
```
Comment (eaidova) [inline models/public/hrnetv2/hrnetv2.md:8]: you have input specification below, no need to include it into common description
```suggestion
and pretrained on ADE20k dataset. 
```

File: models/public/hrnetv2/hrnetv2.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+Image, name - `image`,  shape - `[1x3x320x320]`, format is `[BxCxHxW]`, where:
+
+- `B` - batch size
+- `H` - height
+- `W` - width
+- `C` - channel
+
+Channel order is `RGB`.
+
```
Comment (eaidova) [inline models/public/hrnetv2/hrnetv2.md:43]: please add info about mean and scale values

File: models/public/hrnetv2/hrnetv2.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+### Converted Model
+
+Image, name - `input.1`,  shape - `[1x3x320x320]`, format is `[BxCxHxW]`, where:
+
+- `B` - batch size
+- `C` - channel
+- `H` - height
+- `W` - width
+
+Channel order is `RGB`.
```
Comment (eaidova) [inline models/public/hrnetv2/hrnetv2.md:53]: why not BGR?
in OMZ we prefer to have models with BGR input, because all demos written for this case

File: models/public/hrnetv2/model.yml [hunk 9]
```diff
@@ -0,0 +1,71 @@
+# Copyright (c) 2019 Intel Corporation
```
Comment (eaidova) [inline models/public/hrnetv2/model.yml:1]:
```suggestion
# Copyright (c) 2020 Intel Corporation
```

File: models/public/hrnetv2/model.yml [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+  and pretrained on ADE20k dataset. The model input is a blob that consists of a
+  single image of [1, 3, 320, 320] in RGB order.
+  For details about implementation of model, check out the Semantic Segmentation on MIT ADE20K dataset in PyTorch
+  <https://github.com/CSAILVision/semantic-segmentation-pytorch>.
+task_type: semantic_segmentation
+files:
+  - name: model/official_hrnet.py
+    size: 16557
+    sha256: b5a3df511c3ba6441f6a2d61df65feed50222875353ebc50535d3eb67877e35c
+    source: https://raw.githubusercontent.com/alexmasterdi/semantic-segmentation-pytorch/master/mit_semseg/models/official_hrnet.py
```
Comment (eaidova) [inline models/public/hrnetv2/model.yml:28]: not sure that it is good idea to download modified code from your repo.
@eizamaliev what do you think about it?
Possibly better to upload converted onnx model on download.01.org and here provide instruction how code should be modified for it.

File: tools/accuracy_checker/accuracy_checker/postprocessor/shift.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
+"""
+Copyright (c) 2020 Intel Corporation
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/shift.py:2]:
```suggestion
Copyright (c) 2018-2020 Intel Corporation
```

File: tools/accuracy_checker/accuracy_checker/postprocessor/shift.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    
+    def configure(self):
+        self.shift_value = self.config.get('shift_value')
+    
+    def process_image(self, annotation, prediction):
+
+        for annotation_ in annotation:
+            mask = annotation_.mask.astype(int)
+            update_mask = mask + self.shift_value
+            annotation_.mask = update_mask.astype(np.int16)
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/shift.py:51]: why int16?

File: tools/downloader/pytorch_to_onnx.py [hunk 13]
```diff
@@ -24,6 +24,13 @@ def positive_int_arg(values):
     return result
 
 
+def prepare_model_params(module, model_params):
+    for key in model_params:
+        func = model_params[key].split(".", 1)[0]
+        if hasattr(module, func):
+            model_params[key] = eval("module." + model_params[key])
```
Comment (eaidova) [inline tools/downloader/pytorch_to_onnx.py:31]: could you please describe for what do you use parameters preparation and why simple eval is not enough in your case

File: tools/downloader/pytorch_to_onnx.py [hunk 14]
```diff
@@ -58,10 +65,14 @@ def parse_args():
                         help='Space separated names of the output layers')
     parser.add_argument('--model-param', type=model_parameter, default=[], action='append',
                         help='Pair "name"="value" of model constructor parameter')
+    parser.add_argument('--onnx-version', type=int, default=9,
+                        help='Version of symbolic opset')
+    parser.add_argument('--prepare-params', type=bool, default=False,
```
Comment (eaidova) [inline tools/downloader/pytorch_to_onnx.py:70]: not clear argument name from my view, could you please give more descriptive name for it?

**Comment by eaidova** [comment]
Jenkins ok to test

**Comment by eaidova** [comment]
@alexmasterdi, you have a broken link in models/public/index.md
please fix:
```
15:08:50  ERROR: 1 dead links found!
15:08:50  [✖] ./hrnetv2/hrnetv2.md → Status: 400
15:08:50  [Pipeline] echo
15:08:50  [Error] Found broken links in public/index.md
```

File: tools/accuracy_checker/accuracy_checker/postprocessor/shift.py (hunk 12 above)
Comment (alexmasterdi) [reply tools/accuracy_checker/accuracy_checker/postprocessor/shift.py:51]: Range int8 not big enough to cover all classes. Example: ADE20k with 150(151) classes.

File: models/public/hrnetv2/model.yml (hunk 10 above)
Comment (alexmasterdi) [reply models/public/hrnetv2/model.yml]: At this moment model downloaded from original repo. Also model success converted to IR format, so we don't need to upload converted onnx model on download.01.org.

**Comment by eaidova** [comment]
Jenkins please retry a build

File: tools/downloader/pytorch_to_onnx.py (hunk 13 above)
Comment (alexmasterdi) [reply tools/downloader/pytorch_to_onnx.py]: because function to construct model parameters with simple eval is called before inclusion necessary module.

Comment (eaidova) [reply tools/downloader/pytorch_to_onnx.py]: can we just move inclusion necessary modules before parameters evaluation?
If not, probably this situation should be detected automatically, on my view, without additional user parameters specification. The reason is user is not obligated to know internal mechanics of argument parsing, so now it looks like workaround (if usual loading does not work, then enable this parameter)

File: tools/accuracy_checker/accuracy_checker/postprocessor/shift.py (hunk 12 above)
Comment (eaidova) [reply tools/accuracy_checker/accuracy_checker/postprocessor/shift.py:50]: int8, yes, usually in segmentation used uint8 (0, 255), which enough should be enough (if of course you do not use negative values). int16, can be a problem for some specific postprocessing like resizing. but my question was more about astype uasge 2 line above with int and this, why they are different?

Comment (alexmasterdi) [reply tools/accuracy_checker/accuracy_checker/postprocessor/shift.py:50]: Created this operation by analogy with encode_segmentation_mask, where used int for reading mask and int8 for result mask. But in this case we can't use int8 or uint8, because max in int8 < 151 and if use shift with negative value (example: 0 - 'null' label in ADE20k and shift -1 to annotation) and use uint8, we get error. So I am liable to use int16 as minimal  suitable type. But if this is not so necessary, i can delete astype and use usually int.

File: tools/downloader/pytorch_to_onnx.py (hunk 13 above)
Comment (alexmasterdi) [reply tools/downloader/pytorch_to_onnx.py]: Delete this additional parameter and add eval to postprocessing in model.yml

File: tools/accuracy_checker/accuracy_checker/postprocessor/shift.py (hunk 12 above)
Comment (eaidova) [reply tools/accuracy_checker/accuracy_checker/postprocessor/shift.py:50]: then ok

**Review by eaidova — APPROVED** [review]

File: tools/downloader/pytorch_to_onnx.py [hunk 15]
```diff
@@ -58,6 +58,8 @@ def parse_args():
                         help='Space separated names of the output layers')
     parser.add_argument('--model-param', type=model_parameter, default=[], action='append',
                         help='Pair "name"="value" of model constructor parameter')
+    parser.add_argument('--onnx-version', type=int, default=9,
+                        help='Version of symbolic opset')
```
Comment (eizamaliev) [inline tools/downloader/pytorch_to_onnx.py:62]: You should set default to `11`  (see, [PR](https://github.com/openvinotoolkit/open_model_zoo/pull/1638) for details.
cc @IRDonch

**Review by eizamaliev — CHANGES_REQUESTED** [review]

**Comment by eaidova** [comment]
Possibly, It makes sense add task in model name (we have request for HRNetB2 for pose estimation also) and dataset (in HRNetV2 repo Cityscapes, Pascal Context and LIP dataset trained models are also represented https://github.com/HRNet/HRNet-Semantic-Segmentation)

File: models/public/hrnet-v2-c1/accuracy-check.yaml [hunk 16]
```diff
@@ -0,0 +1,60 @@
+models:
+  - name: hrnetv2_pytorch
```
Comment (eaidova) [inline models/public/hrnet-v2-c1/accuracy-check.yaml:2]: name different from model name in OMZ

File: models/public/hrnet-v2-c1/accuracy-check.yaml [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+    launchers:
+      - framework: onnx_runtime
+        device: CPU
```
Comment (eaidova) [inline models/public/hrnet-v2-c1/accuracy-check.yaml:6]: please remove device

File: models/public/semantic-segmentation-hrnet-v2-c1/accuracy-check.yaml [hunk 18]
```diff
@@ -0,0 +1,59 @@
+models:
```
Comment (IRDonch) [inline models/public/semantic-segmentation-hrnet-v2-c1/accuracy-check.yaml:1]: The name needs to be `accuracy-check.yml`, not `accuracy-check.yaml`.

File: models/public/semantic-segmentation-hrnet-v2-c1/model.yml [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    pattern: '= net_enc'
+    replacement: '= eval(net_enc)'
+  - $type: regex_replace
+    file: mit_semseg/models/models.py
+    pattern: '= net_dec'
+    replacement: '= eval(net_dec)'
+conversion_to_onnx_args:
+  - --model-path=$dl_dir
+  - --model-name=SegmentationModule
+  - --weights=
```
Comment (IRDonch) [inline models/public/semantic-segmentation-hrnet-v2-c1/model.yml:117]: This looks incomplete. I think you should omit this argument and mark the corresponding parameter as optional in the code.

**Review by IRDonch — CHANGES_REQUESTED** [review]

File: models/public/semantic-segmentation-hrnet-v2-c1/semantic-segmentation-hrnet-v2-c1.md [hunk 20]
```diff
@@ -0,0 +1,110 @@
+# semantic-segmentation-hrnet-v2-c1
```
Comment (eizamaliev) [inline models/public/semantic-segmentation-hrnet-v2-c1/semantic-segmentation-hrnet-v2-c1.md:1]: Why did you change name of the model? The `hrnet-v2-c1` was pretty good.

Comment (eaidova) [reply models/public/semantic-segmentation-hrnet-v2-c1/semantic-segmentation-hrnet-v2-c1.md:1]: It was my comment to change, but I can not find it now.
We have potential request for adding hrnet pose estimation too

Comment (eaidova) [reply models/public/semantic-segmentation-hrnet-v2-c1/semantic-segmentation-hrnet-v2-c1.md:1]: my opinion, better <model_name>-<task>
```suggestion
# hrnet-v2-c1-segmentation
```

File: models/public/hrnet-v2-c1-segmentation/hrnet-v2-c1-segmentation.md [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+- `C` - predicted probabilities for each class in [0, 1] range
+- `H` - height
+- `W` - width
+
+### Converted Model
+
+Semantic-segmentation mask according to ADE20k classes, name - `softmax`,  shape - `1,150,320,320`, output data format is `B,C,H,W` where:
+
+- `B` - batch size
+- `C` - predicted probabilities for each class in [0, 1] range
```
Comment (eaidova) [inline models/public/hrnet-v2-c1-segmentation/hrnet-v2-c1-segmentation.md:71]: possibly it makes sense to add some words about labels shift in model description

**Comment by eaidova** [comment]
Jenkins please retry a build

File: models/public/hrnet-v2-c1-segmentation/model.yml [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+    file: mit_semseg/models/models.py
+    pattern: '= net_dec'
+    replacement: '= eval(net_dec)'
+conversion_to_onnx_args:
+  - --model-path=$dl_dir
+  - --model-name=SegmentationModule
+  - --import-module=mit_semseg.models
+  - --input-shape=1,3,320,320
+  - --output-file=$conv_dir/hrnet-v2-c1-segmentation.onnx
+  - --model-param=net_enc=r'ModelBuilder.build_encoder("hrnetv2", 720, r"$dl_dir/ckpt/encoder_epoch_30.pth")'
```
Comment (eaidova) [inline models/public/hrnet-v2-c1-segmentation/model.yml:121]: you use raw string here due to eval issue on Windows. Is my understanding correct?
Do you test this aproach on windows machine? Does it really work?

**Review by IRDonch — APPROVED** [review]

File: models/public/hrnet-v2-c1-segmentation/model.yml (hunk 22 above)
Comment (alexmasterdi) [reply models/public/hrnet-v2-c1-segmentation/model.yml:121]: Yes, it does. I test it on Windows, prefix 'r' doesn't let eval to use escape symbols

Comment (eaidova) [reply models/public/hrnet-v2-c1-segmentation/model.yml:121]: great!

File: models/public/hrnet-v2-c1-segmentation/model.yml [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+    sha256: 55a8f41269be28c17e456e5ae1f246bf71354765739dceca1e45becd58cba5e2
+    source: https://raw.githubusercontent.com/CSAILVision/semantic-segmentation-pytorch/9aff40de31ee4b21f18514d31e5d6e4ba056924d/mit_semseg/models/utils.py
+  - name: mit_semseg/__init__.py
+    size: 63
+    sha256: 7e58a465dbf83903f13d94ffce5b3353bcc8eb2154cd568bb405341f8a02ce5e
+    source: https://raw.githubusercontent.com/CSAILVision/semantic-segmentation-pytorch/9aff40de31ee4b21f18514d31e5d6e4ba056924d/mit_semseg/__init__.py
+  - name: ckpt/decoder_epoch_30.pth
+    size: 4780389
+    sha256: 6060764940b620804befaca4df916de019c82c4e96ace00916991e223af37d1e
+    source: http://sceneparsing.csail.mit.edu/model/pytorch/ade20k-hrnetv2-c1/decoder_epoch_30.pth
```
Comment (eaidova) [inline models/public/hrnet-v2-c1-segmentation/model.yml:93]: oops! looks like downloading server is down now... downloading failed with time out error...

**Review by eizamaliev — APPROVED** [review]

File: models/public/hrnet-v2-c1-segmentation/model.yml (hunk 23 above)
Comment (IRDonch) [reply models/public/hrnet-v2-c1-segmentation/model.yml:93]: Not a good sign... let's hope it was just an isolated incident.

**Comment by IRDonch** [comment]
@alexmasterdi Please rebase, there's a merge conflict.

File: models/public/hrnet-v2-c1-segmentation/model.yml (hunk 23 above)
Comment (eaidova) [reply models/public/hrnet-v2-c1-segmentation/model.yml:93]: in the repo they have issues a year ago that time to time their server under maintenance.
Now http://sceneparsing.csail.mit.edu/ is unreachable, so I think it is also connected with some services on server side.

**Comment by eaidova** [comment]
Jenkins please retry a build

File: models/public/hrnet-v2-c1-segmentation/model.yml (hunk 23 above)
Comment (eaidova) [reply models/public/hrnet-v2-c1-segmentation/model.yml:93]: it is alive! :)
